mldatafind/find.py

```diff
@@ -35,61 +39,84 @@ def _data_generator(
     channels: Iterable[str],
     n_workers: int,
     thread: bool,
+    chunk_size: Optional[float] = None,
     **method_kwargs,
 ) -> Iterator:
 
-    memory_limit = MEMORY_LIMIT
-
     if thread:
         executor = ThreadPoolExecutor(n_workers)
     else:
         executor = ProcessPoolExecutor(n_workers)
 
+    logging.info(f"Finding {len(segments)} segments")
+
     with executor as exc:
         # keep track of current memory
         # and number of futures currently running
         current_memory = 0
-        futures = []
+        futures = {}
 
         # while there are still futures or segments to analyze
         while segments or futures:
 
             # submit jobs until memory limit is reached
-            while current_memory < memory_limit and segments:
+            while current_memory < MEMORY_LIMIT and segments:
                 segment = segments.pop()
 
                 duration = segment[1] - segment[0]
 
-                # TODO: memory depends on sample rate;
-                # for strain channels this is typically 16khz,
-                # but unknown for auxiliary channels
-                segment_memory = _calc_memory(len(channels), duration)
+                segment_memory = utils._estimate_memory(
+                    len(channels), duration
+                )
 
                 future = exc.submit(
                     method, channels, *segment, **method_kwargs
                 )
-                futures.append(future)
+                logging.info(
+                    f"Future submitted to query {duration} s of data"
+                    f"and {segment_memory:.2f} GB of memory"
+                )
+                futures[segment_memory] = future
```

Comment:
I don't know how this idea didn't occur to me before, because I think it would avoid the needed for the `dict` `wait` implementation I have in BBHNet, but it might be easier to invert this: `Future` objects are hashable, so they can be used as the _keys_ to `dict`s. So you set `futures[future] = segment_memory`, then later you can do

```python
done, not_done = concurrent.futures.wait(futures.keys())
for f in done:
    memory = futures.pop(f)
    current_memory -= memory
    yield f.result()
```

Reply:
I also think you'll want to submit a new job (assuming you can given the current memory status) as soon as you yield one future, because if the calling process is spending a lot of time processing this chunk of data, and there are several completed futures, it will be a while before this loop terminates and you get back to the submission section, so you could have a lot of lost parallelization

Reply:
I was also thinking the memory limit could be an argument that defaults to whatever is available. It might be the case that a user want's to allocate memory to other memory intensive processes (e.g. waveform generation) at the same time.